File: back/swift/app/delivery/feature/game/controllers/GameController.swift
import Foundation
import Vapor

class GameController {
    
    private let useCases: GameUseCasesB
    private let sResponses: GameResponses
    
    init(gameUseCases: GameUseCasesB, gameResponses: GameResponses){
        self.useCases = gameUseCases
        self.sResponses = gameResponses
    }
    
    func startGame(_ req: Request) -> Response {

        switch useCases.initializeGame() {
            case .success(let data):
                return handleResult(data: data, successHandler: self.sResponses.successR)
            case .failure(let error):
                return handleError(error: error, errorHandler: self.sResponses.serverErrR)
            case .notFound:
                return self.sResponses.clientErrR(data: .json(JSONData("Not found")), statusCode: .BAD_REQUEST)
        }

    }

    func restartGame(_ req: Request) -> Response {

        do {

            let gameData = try req.content.decode(GameIDData.self)

            if gameData.gameID.isEmpty {
                return self.sResponses.clientErrR(data: .json(JSONData("Game ID cannot be empty")), statusCode: .BAD_REQUEST)
            }

            switch useCases.resetGame(gameID: gameData.gameID) {
                case .success(let data):
                    return handleResult(data: data, successHandler: self.sResponses.successR)
                case .failure(let error):
                    return handleError(error: error, errorHandler: self.sResponses.serverErrR)
                case .notFound:
                    return self.sResponses.clientErrR(data: .json(JSONData("Not found")), statusCode: .BAD_REQUEST)
            }

        } catch {
            return self.sResponses.clientErrR(data: .json(JSONData("Error decoding request data: \(error.localizedDescription)")), statusCode: .BAD_REQUEST)
        }


    }

    func makeMove(_ req: Request) -> Response  {

        do {

            let gameData = try req.content.decode(GameData.self)

            if gameData.gameID.isEmpty {
                return self.sResponses.clientErrR(data: .json(JSONData("Game ID cannot be empty")), statusCode: .BAD_REQUEST)
            }

//            guard let position = gameData.position else {
//                return self.sResponses.clientErrR(data: .json(JSONData("Position cannot be nil")), statusCode: .BAD_REQUEST)
//            }
//
//            guard let playerSymbol = gameData.playerSymbol else {
//                return self.sResponses.clientErrR(data: .json(JSONData("Player symbol cannot be nil")), statusCode: .BAD_REQUEST)
//            }

            let player = PlayerType(symbol: gameData.playerSymbol)

            switch useCases.makeMove(gameID: gameData.gameID, position: gameData.position, player: player) {
                case .success(let data):
                    return handleResult(data: data, successHandler: self.sResponses.successR)
                case .failure(let error):
                    return handleError(error: error, errorHandler: self.sResponses.serverErrR)
                case .notFound:
                    return self.sResponses.clientErrR(data: .json(JSONData("Not found")), statusCode: .BAD_REQUEST)
            }

        }  catch {
        return self.sResponses.clientErrR(data: .json(JSONData("Error decoding request data: \(error.localizedDescription)")), statusCode: .BAD_REQUEST)
        }

    }

    private func handleResult<T: Encodable>(data: T, successHandler: (_ data: SData, _ statusCode: Status.Success) -> Response) -> Response {
        do {
            let jsonData = try JSONEncoder().encode(data)
            return successHandler(.json(JSONData(jsonData)), .OK)
        } catch {
            let errorMessage = "Error encoding data to JSON: \(error)"
            return self.sResponses.serverErrR(data: .json(JSONData(errorMessage)), statusCode: .INTERNAL_SERVER_ERROR)
        }
    }

    private func handleError(error: Error, errorHandler: (_ data: SData, _ statusCode: Status.ServerError) -> Response) -> Response {
        let errorMessage = "Something went wrong: \(error.localizedDescription)"
        return errorHandler(.json(JSONData(errorMessage)), .INTERNAL_SERVER_ERROR)
    }

}

struct GameIDData: Content, Codable {
    let gameID: String
}

struct GameData: Content, Codable {
    let gameID: String
    let position: CellPosition
    let playerSymbol: CellType
}

File: back/kotlin/app/delivery/feature/game/controllers/GameController.kt
package delivery.feature.game.controllers

import io.ktor.server.*
import io.ktor.http.*
import io.ktor.server.application.*
import io.ktor.server.sessions.*
import io.ktor.server.routing.*
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.encodeToString
import kotlinx.serialization.decodeFromString
import kotlinx.serialization.json.Json
import kotlinx.serialization.*
import io.ktor.server.request.*
import io.ktor.server.request.receiveText
import io.ktor.server.request.receiveParameters
import org.slf4j.Logger;
import org.slf4j.LoggerFactory
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.withContext
import kotlinx.coroutines.launch

class GameController(private val useCases: GameUseCasesB, private val responses: GameResponses) {
    
    companion object {
        private val logger = LoggerFactory.getLogger(GameController::class.java)
        private val scope = CoroutineScopesModule.providesCoroutineScope(DispatchersProvider.provideIODispatcher())
        private val gameHTMLContent = GameHTMLContent()
    }

    suspend fun startGame(call: ApplicationCall) {

            logger.info("Starting Game")

            val result = withContext(scope.coroutineContext) {
                useCases.initializeGame()
            }

            logger.info("Game initialization result: $result")

            if (result is Result.Success<GameType>) {
                val boardHtml = gameHTMLContent.getNewBoard()
                withContext(scope.coroutineContext) {
                    saveResult(call, result.data)
                    logger.info("Result Saved")
                    responses.successR(call = call, data = SData.Html(boardHtml), statusCode = Status.Success.OK)
                }
            } else {
                withContext(scope.coroutineContext) {
                    handleResult(result, call)
                }
            }

    }

    suspend fun restartGame(call: ApplicationCall) {

        logger.info("Restarting Game")
        val savedResult = call.sessions.get<GameSession>()
        val gameID = savedResult?.gameID
        if (gameID == null) {
            responses.serverErrR(call = call, data = SData.Json(emptyMap()), statusCode = Status.ServerError.INTERNAL_SERVER_ERROR)
            return
        }
        val result = withContext(scope.coroutineContext) {
            useCases.resetGame(gameID)
        }
        logger.info("Game restart result: $result")
        if (result is Result.Success<GameType>) {
            withContext(scope.coroutineContext) {
                saveResult(call, result.data)
                val boardHtml = gameHTMLContent.getNewBoard()
                responses.successR(call = call, data = SData.Html(boardHtml), statusCode = Status.Success.OK)
            }
        } else {
            withContext(scope.coroutineContext) {
                handleResult(result, call)
            }
        }

    }

    suspend fun makeMove(call: ApplicationCall) {

            logger.info("Making Move")
            val positionData = call.receiveParameters()["position"]

            val savedResult = call.sessions.get<GameSession>()
            val gameID = savedResult?.gameID
            val currentPlayer = savedResult?.currentPlayer

            if (gameID == null || currentPlayer == null || positionData == null) {
                responses.serverErrR(call = call, data = SData.Json(emptyMap()), statusCode = Status.ServerError.INTERNAL_SERVER_ERROR)
                return
            }

            val position = CellPosition.valueOf(positionData)

            val result = withContext(scope.coroutineContext) {
                useCases.makeMove(gameID, position, currentPlayer)
            }

            logger.info("Game Move result: $result")

            if (result is Result.Success<GameType>) {

                val newTitle = when {
                    result.data.gameState == GameState.WON && result.data.winner != null -> {
                        if (result.data.winner.symbol == CellType.X) GameTitle.PLAYER_X_WON else GameTitle.PLAYER_O_WON
                    }
                    result.data.gameState == GameState.DRAW -> GameTitle.DRAW
                    else -> GameTitle.PLAYING
                }

                when (result.data.state) {
                    is State.BoardState -> {
                        logger.info("Board state is an array or a dictionary")
                        responses.serverErrR(call = call, data = SData.Json(emptyMap()), statusCode = Status.ServerError.INTERNAL_SERVER_ERROR)
                        return
                    }
                    is State.MovesState -> {
                        withContext(scope.coroutineContext) {
                            saveResult(call, result.data)
                            val newMove = gameHTMLContent.getBoard(newTitle, result.data.state.moves)
                            responses.successR(call = call, data = SData.Html(newMove), statusCode = Status.Success.OK)
                        }
                    }
                }

            } else {
                withContext(scope.coroutineContext) {
                    handleResult(result, call)
                }
            }

    }

    private suspend fun handleResult(result: Result<GameType>, call: ApplicationCall) {
        when {
//            result is Result.Success<GameType> -> responses.successR(
//                call = call,
//                data = SData.Json(mapOf("data" to result.data)),
//                statusCode = Status.Success.OK
//            )
            result is Result.Error -> responses.serverErrR(
                call = call,
                data = SData.Json(emptyMap()),
                statusCode = Status.ServerError.INTERNAL_SERVER_ERROR
            )
            result is Result.NotFound -> responses.clientErrR(
                call = call,
                data = SData.Json(emptyMap()),
                statusCode = Status.ClientError.REQUEST_TIMEOUT
            )
        }
    }

    private suspend fun saveResult(call: ApplicationCall, result: GameType) {

        val serializedState = Json.encodeToString(result.state)

        call.sessions.set(GameSession(
            gameID = result.id,
            currentPlayer = result.currentPlayer,
            gameState = result.gameState,
            state = serializedState,
            winner = result.winner
        ))
    }

}